pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -218,50 +218,87 @@ func (p *DNSProxy) checkRestored(endpointID uint64, destPort uint16, destIP stri
+	defer p.RUnlock()
+	if len(p.usedServers) > 0 {
+		if _, used := p.usedServers[ip]; !used {
+			return true
+		}
+	}
+	return false
+}
+
 // GetRules creates a fresh copy of EP's DNS rules to be stored
 // for later restoration.
 func (p *DNSProxy) GetRules(endpointID uint16) (restore.DNSRules, error) {
+	// Lock ordering note: Acquiring the IPCache read lock (as LookupIPsBySecID does) while holding
+	// the proxy lock can lead to a deadlock. Avoid this by reading the state from DNSProxy while
+	// holding the read lock, then perform the IPCache lookups.
+	// Note that IPCache state may change in between calls to LookupIPsBySecID.
 	p.RLock()
-	defer p.RUnlock()
 
-	restored := make(restore.DNSRules)
+	type selRegex struct {
+		re *regexp.Regexp
+		cs policy.CachedSelector
+	}
+
+	portToSelRegex := make(map[uint16][]selRegex)
 	for port, entries := range p.allowed[uint64(endpointID)] {
-		var ipRules restore.IPRules
+		var nidRules = make([]selRegex, 0, len(entries))
+		// Copy the entries to avoid racy map accesses after we release the lock. We don't need
+		// constant time access, hence a preallocated slice instead of another map.
 		for cs, regex := range entries {
-			var IPs map[string]struct{}
-			if !cs.IsWildcard() {
-				IPs = make(map[string]struct{})
-				count := 0
-			Loop:
-				for _, nid := range cs.GetSelections() {
-					nidIPs := p.LookupIPsBySecID(nid)
-					for _, ip := range nidIPs {
-						// Skip IPs that are allowed but have never been used,
-						// but only if at least one server has been used so far.
-						if len(p.usedServers) > 0 {
-							if _, used := p.usedServers[ip]; !used {
-								continue
-							}
-						}
-						IPs[ip] = struct{}{}
-						count++
-						if count > p.maxIPsPerRestoredDNSRule {
-							log.WithFields(logrus.Fields{
-								logfields.EndpointID:            endpointID,
-								logfields.Port:                  port,
-								logfields.EndpointLabelSelector: cs,
-								logfields.Limit:                 p.maxIPsPerRestoredDNSRule,
-								logfields.Count:                 len(nidIPs),
-							}).Warning("Too many IPs for a DNS rule, skipping the rest")
-							break Loop
-						}
+			nidRules = append(nidRules, selRegex{cs: cs, re: regex})
+		}
+		portToSelRegex[port] = nidRules
+	}
+
+	maxIPsPerRestoredDNSRule := p.maxIPsPerRestoredDNSRule
+
+	// We've read what we need from the proxy. The following IPCache lookups _must_ occur outside of
+	// the critical section.
+	p.RUnlock()
+
+	restored := make(restore.DNSRules)
+	for port, selRegexes := range portToSelRegex {
+		var ipRules restore.IPRules
+		for _, selRegex := range selRegexes {
+			if selRegex.cs.IsWildcard() {
+				ipRules = append(ipRules, asIPRule(selRegex.re, nil))
+				continue
+			}
+			ips := make(map[string]struct{})
+			count := 0
+			nids := selRegex.cs.GetSelections()
+		Loop:
+			for _, nid := range nids {
+				nidIPs := p.LookupIPsBySecID(nid)
+				for _, ip := range nidIPs {
+					if p.skipIPInRestoration(ip) {
```

Comment:
Does it make sense to acquire the proxy read-lock here in a loop like this or should we move the lock outside of the lock and make `skipIPInRestoration()` -> `skipIPInRestorationRLocked()`?

Reply:
Good call! I moved it out of the `for _, ip` loop. Moving it out of the `for _, nid` loop is a bit more tricky, so I'll leave this as is for now.